src/systems/filecoin_vm/interpreter/vm_interpreter.go

```diff
@@ -14,172 +15,255 @@ import (
+			outTree, receipt, minerPenaltyCurr = vmi.ApplyMessage(outTree, m, onChainMessageLen, blk.Miner())
+			minerPenaltyTotal += minerPenaltyCurr
+			receipts = append(receipts, receipt)
 			seenMsgs[_msgCID(m)] = struct{}{}
 		}
+
 		// Process SECP messages from the block.
 		for _, sm := range blk.SECPMessages() {
 			m := sm.Message()
 			_, found := seenMsgs[_msgCID(m)]
 			if found {
 				continue
 			}
-			outTree, r = vmi.ApplyMessage(outTree, m, blk.Miner())
-			receipts = append(receipts, r)
+			onChainMessageLen := len(msg.Serialize_SignedMessage(sm))
+			outTree, receipt, minerPenaltyCurr = vmi.ApplyMessage(outTree, m, onChainMessageLen, blk.Miner())
+			minerPenaltyTotal += minerPenaltyCurr
+			receipts = append(receipts, receipt)
 			seenMsgs[_msgCID(m)] = struct{}{}
 		}
+
+		// Pay block reward.
+		rewardMessage := _makeBlockRewardMessage(outTree, minerOwner, minerPenaltyTotal)
+		outTree = _applyMessageBuiltinAssert(outTree, rewardMessage, blk.Miner())
 	}
 
 	// Invoke cron tick, attributing it to the miner of the first block.
-	cronSender := msgs.Blocks()[0].Miner()
-	cron := _makeCronTickMessage(outTree, cronSender)
-	outTree, r = vmi.ApplyMessage(outTree, cron, cronSender)
-	if r.ExitCode() != exitcode.OK() {
-		panic("cron tick failed")
-	}
+
+	TODO()
+	// TODO: miners shouldn't be able to trigger cron by sending messages;
+	// use ControlActor instead (https://github.com/filecoin-project/specs/issues/665)
+
+	firstMiner := msgs.Blocks()[0].Miner()
+	cronMessage := _makeCronTickMessage(outTree, firstMiner)
+	outTree = _applyMessageBuiltinAssert(outTree, cronMessage, firstMiner)
 
 	return
 }
 
-func (vmi *VMInterpreter_I) ApplyMessage(inTree st.StateTree, message msg.UnsignedMessage, minerAddr addr.Address) (
-	st.StateTree, vmr.MessageReceipt) {
+func (vmi *VMInterpreter_I) ApplyMessage(
+	inTree st.StateTree, message msg.UnsignedMessage, onChainMessageSize int, minerAddr addr.Address) (
+	retTree st.StateTree, retReceipt vmr.MessageReceipt, retMinerPenalty actor.TokenAmount) {
 
-	compTree := inTree
-	var outTree st.StateTree
-	var toActor actor.ActorState
-	var err error
+	vmiGasRemaining := message.GasLimit()
+	vmiGasUsed := msg.GasAmount_Zero()
+
+	_applyReturn := func(
+		tree st.StateTree, invocOutput vmr.InvocOutput, exitCode exitcode.ExitCode,
+		senderResolveSpec SenderResolveSpec) {
+
+		switch senderResolveSpec {
+		case SenderResolveSpec_OK:
+			// In this case, the sender is valid and has already transferred funds to the burnt funds actor
+			// sufficient for the gas limit. Thus, we may refund the unused gas funds to the sender here.
+			Assert(!message.GasLimit().LessThan(vmiGasUsed))
+			Assert(message.GasLimit().Equals(vmiGasUsed.Add(vmiGasRemaining)))
+			vmiGasRemainingFIL := _gasToFIL(vmiGasRemaining, message.GasPrice())
+			tree = _withTransferFundsAssert(tree, addr.BurntFundsActorAddr, message.From(), vmiGasRemainingFIL)
+			retMinerPenalty = actor.TokenAmount(0)
```

Comment:
The miner hasn't been paid!

Reply:
Yes, that's a bug -- just fixed. (I had missed this earlier since it's still very counterintuitive to me that a miner causes the network to incur costs here, and then gets paid for doing so!)